opengovsg/postmangovsg
# feat: add lambda to process sns sms publish logs

_This PR is part of the SNS SMS fallback feature. This is broken out to a separate PR to make it easier to review_

## Problem

We needed a way to get the final delivery status of SMS sent using SNS when the fallback is activated. The final delivery status is only published to CloudWatch as logs. 

## Solution

**Features**:
- [Enable logging](https://docs.aws.amazon.com/sns/latest/dg/sms_stats_cloudwatch.html) for SNS SMS to CloudWatch 
- Create subscription filter for logs to trigger lambda
- Add lambda to process log events and update `sms_messages` table with final delivery status

## Tests
- Setup and configure subscription filters for CloudWatch logs using instructions in the [incident response guide](https://docs.google.com/document/d/1Trcqr17UZBjBoCOUKuzkk7vbbFR99QkqRobhtQ-sVIc/edit#heading=h.eoj9sln41yja).
- Send an SMS campaign
- Download delivery report and ensure that the status is correct

## Deploy Notes
**Create new database user**
```sql
CREATE USER serverless_log_sns_sms_iam;
GRANT rds_iam TO serverless_log_sns_sms_iam;
GRANT SELECT, UPDATE ON sms_messages TO serverless_log_sns_sms_iam;
```

**New environment variables**
- `PRODUCTION_LOG_SNS_SMS_ROLE`
- `PRODUCTION_LOG_SNS_SMS_DB_URI`
- `STAGING_LOG_SNS_SMS_ROLE`
- `STAGING_LOG_SNS_SMS_DB_URI`

## TODOs
- [x] Create staging database user and IAM policy for lambda 
- [x] Add deploy block to Travis
- [x] Test on staging
- [x] Create production IAM policy and role for lambda
